test/test_torch.py

```diff
@@ -7023,67 +7023,75 @@ def test_logical_not_out(self, device, dtypes):
-                # Skip only if either dtype or other_dtype is bfloat16, and promotion is unsupported.
-                if (dtype == torch.bfloat16) != (other_dtype == torch.bfloat16):
-                    non_bfloat16_dtype = dtype if dtype != torch.bfloat16 else other_dtype
-                    if non_bfloat16_dtype.is_complex or non_bfloat16_dtype == torch.float16:
-                        with self.assertRaises(RuntimeError):
-                            getattr(a, op)(b)
-                        continue
+    def _test_logical(self, device, dtypes, op, a_, b_, expected_res_):
+        expected_res = torch.tensor(expected_res_, dtype=dtypes[0], device=device)
+        a = torch.tensor(a_, dtype=dtypes[0], device=device)
+        b = torch.tensor(b_, dtype=dtypes[1], device=device)
+
+        # TODO Remove this skipping after bfloat16 can be handled nicely with other dtypes for complex and float16.
+        # Skip only if either dtype or other_dtype is bfloat16, and promotion is unsupported.
+        if (dtypes[0] == torch.bfloat16) != (dtypes[1] == torch.bfloat16):
+            non_bfloat16_dtype = dtypes[0] if dtypes[0] != torch.bfloat16 else dtypes[1]
+            if non_bfloat16_dtype.is_complex or non_bfloat16_dtype == torch.float16:
+                with self.assertRaises(RuntimeError):
+                    getattr(a, op)(b)
+                return
 
-                # Skip bfloat16 on CUDA. Remove this after bfloat16 is supported on CUDA.
-                # After type promotion of bfloat16 is supported, some bfloat16 logical operation will go through on
-                # CUDA as long as the two tensors are promoted to a supported type.
-                # TODO: Remove this once logical operators are improved to take care of bfloat16.
-                if self.device_type == 'cuda' and torch.bfloat16 in (dtype, other_dtype):
-                    if torch.promote_types(dtype, other_dtype) == torch.bfloat16:
-                        with self.assertRaises(RuntimeError):
-                            getattr(a, op)(b)
-                        continue
+        # Skip bfloat16 on CUDA. Remove this after bfloat16 is supported on CUDA.
+        # After type promotion of bfloat16 is supported, some bfloat16 logical operation will go through on
+        # CUDA as long as the two tensors are promoted to a supported type.
+        # TODO: Remove this once logical operators are improved to take care of bfloat16.
+        if self.device_type == 'cuda' and torch.bfloat16 in dtypes:
+            if torch.promote_types(dtypes[0], dtypes[1]) == torch.bfloat16:
+                with self.assertRaises(RuntimeError):
+                    getattr(a, op)(b)
+                return
 
-                if dtype.is_complex or other_dtype.is_complex:
-                    with self.assertRaises(RuntimeError):
-                        getattr(a, op)(b)
-                    continue
+        if dtypes[0].is_complex or dtypes[1].is_complex:
```

Comment:
Nit: I don't think this case can actually be hit because of the preceding checks for bfloat16 x complex

Reply:
Oh you're right, sorry. I thought this was nested under another check and it's not. My mistake.